File: indexer/indexer.go
package indexer

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"sync/atomic"

	"github.com/AkashGoenka/coldstart/graph"
	"github.com/AkashGoenka/coldstart/parser"
)

// Config holds indexer settings.
type Config struct {
	RootDir    string
	Extensions map[string]bool // e.g. {".ts": true, ".tsx": true}
	Exclude    map[string]bool // dir names to skip e.g. {"node_modules": true}
	Workers    int             // goroutine pool size
	HasReact   bool            // true if package.json lists react as a dependency
}

// DefaultConfig returns a sensible default for TS/JS/GQL codebases.
func DefaultConfig(root string) *Config {
	return &Config{
		RootDir: root,
		Extensions: map[string]bool{
			".ts": true, ".tsx": true,
			".js": true, ".jsx": true,
			".mjs": true, ".cjs": true,
			".graphql": true, ".gql": true,
		},
		Exclude: map[string]bool{
			"node_modules": true,
			"dist":         true,
			"build":        true,
			".git":         true,
			".next":        true,
			".turbo":       true,
			"coverage":     true,
			"__pycache__":  true,
		},
		Workers: 16,
	}
}

// Stats holds indexing statistics.
type Stats struct {
	FilesScanned  int64
	FilesIndexed  int64
	FilesSkipped  int64
	TotalTokens   int64
	EdgesResolved int64
}

// Run walks the codebase concurrently and builds the dependency graph.
func Run(cfg *Config) (*graph.Graph, *Stats, error) {
	g := graph.New()
	stats := &Stats{}

	cfg.HasReact = detectReact(cfg.RootDir)

	// ── Phase 1: Collect all file paths ──────────────────────────────────────
	var filePaths []string
	err := filepath.WalkDir(cfg.RootDir, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return nil // skip unreadable paths
		}
		if d.IsDir() {
			if cfg.Exclude[d.Name()] {
				return filepath.SkipDir
			}
			return nil
		}
		ext := filepath.Ext(path)
		if cfg.Extensions[ext] {
			filePaths = append(filePaths, path)
			atomic.AddInt64(&stats.FilesScanned, 1)
		}
		return nil
	})
	if err != nil {
		return nil, nil, err
	}

	// ── Phase 2: Parse files concurrently ────────────────────────────────────
	jobs := make(chan string, len(filePaths))
	for _, p := range filePaths {
		jobs <- p
	}
	close(jobs)

	var wg sync.WaitGroup
	for i := 0; i < cfg.Workers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for absPath := range jobs {
				relPath, _ := filepath.Rel(cfg.RootDir, absPath)
				relPath = filepath.ToSlash(relPath) // normalize to forward slashes

				var node *graph.Node
				var err error

				ext := filepath.Ext(absPath)
				if ext == ".graphql" || ext == ".gql" {
					node, err = parser.ParseGQLFile(absPath, relPath)
				} else {
					node, err = parser.ParseFile(absPath, relPath, cfg.HasReact)
				}

				if err != nil {
					atomic.AddInt64(&stats.FilesSkipped, 1)
					continue
				}

				g.AddNode(node)
				atomic.AddInt64(&stats.FilesIndexed, 1)
				atomic.AddInt64(&stats.TotalTokens, int64(node.TokenEstimate))
			}
		}()
	}
	wg.Wait()

	// ── Phase 3: Resolve edges (import specifiers → actual file IDs) ─────────
	resolveEdges(g, cfg.RootDir, stats)

	return g, stats, nil
}

// resolveEdges converts raw import strings into graph edges.
// It resolves relative imports to actual file IDs in the graph.
func resolveEdges(g *graph.Graph, rootDir string, stats *Stats) {
	// Build a lookup set of all known node IDs for fast resolution
	knownIDs := make(map[string]bool, len(g.Nodes))
	for id := range g.Nodes {
		knownIDs[id] = true
	}

	for id, node := range g.Nodes {
		for _, imp := range node.Imports {
			// Only resolve relative imports (starting with . or ..)
			if !strings.HasPrefix(imp, ".") {
				continue
			}

			resolved := resolveRelativeImport(id, imp, knownIDs)
			if resolved != "" {
				g.AddEdge(id, resolved, "imports")
				atomic.AddInt64(&stats.EdgesResolved, 1)
			}
		}
	}
}

// detectReact reads {root}/package.json and returns true if react is listed
// as a dependency, devDependency, or peerDependency.
func detectReact(root string) bool {
	data, err := os.ReadFile(filepath.Join(root, "package.json"))
	if err != nil {
		return false
	}
	var pkg struct {
		Dependencies    map[string]string `json:"dependencies"`
		DevDependencies map[string]string `json:"devDependencies"`
		PeerDependencies map[string]string `json:"peerDependencies"`
	}
	if err := json.Unmarshal(data, &pkg); err != nil {
		return false
	}
	_, inDeps := pkg.Dependencies["react"]
	_, inDev := pkg.DevDependencies["react"]
	_, inPeer := pkg.PeerDependencies["react"]
	return inDeps || inDev || inPeer
}

// resolveRelativeImport tries to find the actual file ID for a relative import.
// Handles: ./foo → foo.ts, ./foo → foo/index.ts, etc.
func resolveRelativeImport(fromID, importSpec string, knownIDs map[string]bool) string {
	fromDir := filepath.ToSlash(filepath.Dir(fromID))
	base := filepath.ToSlash(filepath.Join(fromDir, importSpec))

	// Try direct extensions
	extensions := []string{".ts", ".tsx", ".js", ".jsx", ".mjs", ".cjs"}
	for _, ext := range extensions {
		candidate := base + ext
		if knownIDs[candidate] {
			return candidate
		}
	}

	// Try index files (e.g. ./auth → auth/index.ts)
	for _, ext := range extensions {
		candidate := base + "/index" + ext
		if knownIDs[candidate] {
			return candidate
		}
	}

	return "" // unresolved (external or missing)
}

File: indexer-node/src/indexer.js
/**
 * indexer.js — Core indexer: walks directories, farms out parsing to a worker
 * pool, then builds the graph and serializes the output object.
 */

import { readdirSync, statSync, readFileSync } from 'node:fs';
import { join, extname, relative } from 'node:path';
import { cpus } from 'node:os';
import { Worker } from 'node:worker_threads';
import { Graph } from './graph.js';

const WORKER_PATH = new URL('./worker.js', import.meta.url).pathname;

const EXTENSIONS = new Set(['.ts', '.tsx', '.js', '.jsx', '.mjs', '.cjs', '.graphql', '.gql']);

const DEFAULT_EXCLUDE = new Set([
  'node_modules', 'dist', 'build', '.git', '.next', '.turbo', 'coverage', '__pycache__',
]);

const HOT_NODE_THRESHOLD = 5;
const INDEXER_VERSION = '1.0.0-node';

/**
 * Run the full indexing pipeline.
 *
 * @param {object} opts
 * @param {string} opts.rootDir        Absolute path to the project root
 * @param {string[]} [opts.extraExclude]  Additional dir names to exclude
 * @param {boolean} [opts.withArchitecture] Include entrypoint/depth/critical-path analysis
 * @returns {Promise<{output: object, stats: object, graph: Graph}>}
 */
export async function runIndexer({ rootDir, extraExclude = [], includePaths = [], withArchitecture = false }) {
  const excludeDirs = new Set([...DEFAULT_EXCLUDE, ...extraExclude]);
  const hasReact = detectReact(rootDir);

  // ── Phase 1: Walk and collect file paths ───────────────────────────────
  const stats = {
    FilesScanned: 0,
    FilesIndexed: 0,
    FilesSkipped: 0,
    TotalTokens: 0,
    EdgesResolved: 0,
  };

  const filePaths = [];
  const roots = includePaths.length > 0 ? includePaths : [rootDir];
  for (const root of roots) {
    walkDir(root, excludeDirs, filePaths);
  }
  stats.FilesScanned = filePaths.length;

  // ── Phase 2: Parse files in worker pool ────────────────────────────────
  const graph = new Graph();
  const numWorkers = Math.max(1, cpus().length);
  const nodes = await parseWithWorkerPool(filePaths, rootDir, hasReact, numWorkers, stats);

  for (const node of nodes) {
    graph.addNode(node);
  }

  // ── Phase 3: Resolve edges ─────────────────────────────────────────────
  graph.resolveEdges(rootDir, stats);

  // ── Phase 3b: Optional architecture tracing ────────────────────────────
  if (withArchitecture) {
    graph.computeDepths();
  }

  // ── Phase 4: Build output ──────────────────────────────────────────────
  const sortedNodes = [...graph.nodes.values()].sort((a, b) => a.id < b.id ? -1 : a.id > b.id ? 1 : 0);

  // Normalize node fields to match Go output schema exactly
  const outputNodes = sortedNodes.map(n => normalizeNode(n));

  let architecture_layers;
  let critical_paths = [];
  if (withArchitecture) {
    // Architecture layers — global grouped view by structural role
    const roleToKey = { router: 'routers', service: 'services', repository: 'repositories', middleware: 'middleware' };
    architecture_layers = { routers: [], services: [], repositories: [], middleware: [] };
    for (const node of graph.nodes.values()) {
      const key = roleToKey[node.architectural_role];
      if (key) architecture_layers[key].push(node.id);
    }
    for (const key of Object.keys(architecture_layers)) architecture_layers[key].sort();

    // Critical paths — per-entry-point role chains
    critical_paths = graph.computeCriticalPaths();
  }

  // Hot nodes — sorted by descending dependents count
  const hotNodeIDs = graph.hotNodes(HOT_NODE_THRESHOLD);
  const hotNodes = hotNodeIDs
    .map(id => {
      const node = graph.nodes.get(id);
      return {
        id,
        dependents: graph.dependents(id).length,
        domain: node?.domain ?? '',
      };
    })
    .sort((a, b) => b.dependents - a.dependents);

  const cycles = graph.detectCycles();
  const clusters = graph.clusterByDomain();

  // Edges: rename 'kind' -> 'kind' (Go uses 'type' field name but JSON key is 'kind')
  // Actually Go uses json:"type" not "kind" — let's check:
  // graph.go: Type string `json:"type"` — so output key is "type" not "kind"
  // Our graph stores { from, to, kind } — need to rename to match Go
  const outputEdges = graph.edges.map(e => ({ from: e.from, to: e.to, type: e.kind }));

  const output = {
    meta: {
      generated_at:    new Date().toISOString().replace(/\.\d{3}Z$/, 'Z'),
      root_dir:        rootDir,
      total_files:     outputNodes.length,
      total_edges:     outputEdges.length,
      total_tokens:    stats.TotalTokens,
      indexer_version: INDEXER_VERSION,
    },
    nodes:               outputNodes,
    edges:               outputEdges,
    clusters,
    ...(architecture_layers ? { architecture_layers } : {}),
    hot_nodes:           hotNodes,
    cycles:              cycles.length > 0 ? cycles : undefined,
    ...(critical_paths.length > 0 ? { critical_paths } : {}),
    stats,
  };

  return { output, stats, graph };
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Recursively walk a directory, collecting files with supported extensions.
 */
function walkDir(dir, excludeDirs, result) {
  let entries;
  try {
    entries = readdirSync(dir, { withFileTypes: true });
  } catch {
    return; // unreadable dir — skip
  }

  for (const entry of entries) {
    if (entry.isDirectory()) {
      if (!excludeDirs.has(entry.name)) {
        walkDir(join(dir, entry.name), excludeDirs, result);
      }
    } else if (entry.isFile()) {
      const ext = extname(entry.name).toLowerCase();
      if (EXTENSIONS.has(ext)) {
        result.push(join(dir, entry.name));
      }
    }
  }
}

/**
 * Distribute file parsing across a pool of workers.
 */
function parseWithWorkerPool(filePaths, rootDir, hasReact, numWorkers, stats) {
  return new Promise((resolve, reject) => {
    if (filePaths.length === 0) {
      resolve([]);
      return;
    }

    const nodes = [];
    const actualWorkers = Math.min(numWorkers, filePaths.length);
    let jobIndex = 0;
    let finished = 0;

    const workers = [];

    function dispatchNext(worker) {
      if (jobIndex >= filePaths.length) {
        // No more jobs for this worker — terminate it when all are done
        worker.postMessage(null); // sentinel to signal termination
        return;
      }

      const absPath = filePaths[jobIndex++];
      const relPath = relative(rootDir, absPath).replace(/\\/g, '/');
      worker.postMessage({ absPath, relPath, hasReact });
    }

    for (let i = 0; i < actualWorkers; i++) {
      const worker = new Worker(WORKER_PATH);

      worker.on('message', ({ node, error, relPath }) => {
        if (node) {
          nodes.push(node);
          stats.FilesIndexed++;
          stats.TotalTokens += node.token_estimate ?? 0;
        } else {
          stats.FilesSkipped++;
        }
        dispatchNext(worker);
      });

      worker.on('error', (err) => {
        stats.FilesSkipped++;
        dispatchNext(worker);
      });

      worker.on('exit', () => {
        finished++;
        if (finished === actualWorkers) {
          resolve(nodes);
        }
      });

      workers.push(worker);
      dispatchNext(worker);
    }
  });
}

/**
 * Check if root/package.json lists react as a dependency.
 */
function detectReact(rootDir) {
  try {
    const raw = readFileSync(join(rootDir, 'package.json'), 'utf8');
    const pkg = JSON.parse(raw);
    return !!(
      pkg.dependencies?.react ||
      pkg.devDependencies?.react ||
      pkg.peerDependencies?.react
    );
  } catch {
    return false;
  }
}

/**
 * Normalize a node to match the Go output schema exactly.
 * Ensures all required fields are present and optional fields follow the same rules.
 */
function normalizeNode(n) {
  return {
    id:             n.id,
    language:       n.language,
    summary:        n.summary ?? '',
    exports:        n.exports ?? [],
    imports:        n.imports ?? [],
    domain:         n.domain ?? 'misc',
    is_entry_point: n.is_entry_point ?? false,
    line_count:     n.line_count ?? 0,
    token_estimate: n.token_estimate ?? 0,
    hash:           n.hash ?? '',
    // omitempty equivalents
    ...(n.depth !== undefined ? { depth: n.depth } : {}),
    ...(n.reachable !== undefined ? { reachable: n.reachable } : {}),
    ...(n.architectural_role ? { architectural_role: n.architectural_role } : {}),
    ...(n.hook_names && n.hook_names.length > 0 ? { hook_names: n.hook_names } : {}),
    ...(n.jsx_components && n.jsx_components.length > 0 ? { jsx_components: n.jsx_components } : {}),
    ...(n.props_types && n.props_types.length > 0 ? { props_types: n.props_types } : {}),
    ...(n.internal_hooks && n.internal_hooks.length > 0 ? { internal_hooks: n.internal_hooks } : {}),
    ...(n.gql ? { gql: n.gql } : {}),
  };
}
